Add support for client certificates

amoerie opened the pull request:
Fixes #1402
Fixes #903
Fixes #688 

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Add support for using X509 certificate(s) when authenticating over SSL as a client to a DICOM server
- Add DicomServerTlsOptions (this is a parameter in DicomServerFactory methods) and DicomClientTlsOptions (this is exposed as TlsOptions on NetworkStreamCreationOptions, which in turn is exposed on DicomClient as NetworkStreamCreationOptions)

#### Breaking changes proposed in this pull request:
- DicomServerFactory.Create now accepts DicomServerTlsOptions instead of a X509Certificate. 
- The option 'IgnoreSslPolicyErrors' has been removed. For both client and server, you can now configure a RemoteCertificationValidationCallback that always returns true if you still want the same behavior.

@gofal Since this introduces a breaking change, perhaps a minor version bump is warranted? E.g. 5.1.0 instead of 5.0.4

amoerie committed:
Don't do sync over async when creating network streams

amoerie committed:
Expose client SSL certificate options on NetworkStreamCreationOptions

amoerie committed:
Add support for client certificates

amoerie committed:
Request client certificate as server

amoerie committed:
Use instance of X509Certificate2

MBesuden commented on FO-DICOM.Core/Network/DesktopNetworkStream.cs:
```diff
@@ -97,45 +50,97 @@ internal DesktopNetworkStream(string host, int port, bool useTls, bool noDelay,
         /// if <paramref name="ownsTcpClient"/> is true, <paramref name="tcpClient"/> will be disposed altogether on DesktopNetworkStream's disposal.
         /// </remarks>
-        internal DesktopNetworkStream(TcpClient tcpClient, X509Certificate certificate, bool ownsTcpClient = false)
+        internal static async Task<DesktopNetworkStream> CreateAsServerAsync(TcpClient tcpClient, X509Certificate certificate, bool ownsTcpClient, CancellationToken cancellationToken)
```
As client authentication is needed for one of our projects and we would like to enforce client authentication i would like to propose to add a check if the mutual authentication succeeded (which is not enforced by AuthenticateAsServerAsync if no client certificate is given, see [AuthenticateAsServerAsync](https://docs.microsoft.com/en-us/dotnet/api/system.net.security.sslstream.authenticateasserverasync?view=netstandard-2.0#system-net-security-sslstream-authenticateasserverasync(system-security-cryptography-x509certificates-x509certificate-system-boolean-system-security-authentication-sslprotocols-system-boolean)))).

I would like to propose to add an optional check for [SslStream.IsMutuallyAuthenticated](https://docs.microsoft.com/en-us/dotnet/api/system.net.security.sslstream.ismutuallyauthenticated?view=netstandard-2.0):
```cs
internal static async Task<DesktopNetworkStream> CreateAsServerAsync(TcpClient tcpClient, X509Certificate certificate, bool ownsTcpClient, CancellationToken cancellationToken, bool forceClientAuthentication = false)
{
    var localHost = ((IPEndPoint)tcpClient.Client.LocalEndPoint).Address.ToString();
    var localPort = ((IPEndPoint)tcpClient.Client.LocalEndPoint).Port;
    var remoteHost = ((IPEndPoint)tcpClient.Client.RemoteEndPoint).Address.ToString();
    var remotePort = ((IPEndPoint)tcpClient.Client.RemoteEndPoint).Port;

    Stream stream = tcpClient.GetStream();
    if (certificate != null)
    {
        var ssl = new SslStream(stream, false);

        var sslHandshake = Task.Run(() => ssl.AuthenticateAsServerAsync(certificate, true, SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12, false), cancellationToken);
        var sslHandshakeTimeout = Task.Delay(_sslHandshakeTimeout, cancellationToken);

        if (await Task.WhenAny(sslHandshake, sslHandshakeTimeout).ConfigureAwait(false) == sslHandshakeTimeout)
        {
            throw new DicomNetworkException($"SSL server authentication took longer than {_sslHandshakeTimeout.TotalSeconds}s");
        }

        await sslHandshake.ConfigureAwait(false);
        
        if (forceClientAuthentication && !ssl.IsMutuallyAuthenticated)
        {
            throw new DicomNetworkException("SSL client authentication failed");
        }

        stream = ssl;
    }

    return new DesktopNetworkStream(localHost, localPort, remoteHost, remotePort, stream, ownsTcpClient ? tcpClient : null);
}
```

amoerie replied:
Please review my latest changes, I've incorporated your feedback and exposed this as an extra option in DicomServiceOptions.

amoerie committed:
Make GH-1359 test less flaky

One day, we'll find a better way to write these time based tests...

amoerie committed:
Add option 'RequireMutualAuthentication' and use Options pattern to start a network listener

amoerie committed:
Add working tests for (mutual) SSL authentication

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1402

MBesuden commented on FO-DICOM.Core/Network/DicomServiceOptions.cs:
```diff
@@ -75,7 +83,8 @@ public class DicomServiceOptions
                 MaxPDULength = MaxPDULength,
-                MaxPDVsPerPDU = MaxPDVsPerPDU
+                MaxPDVsPerPDU = MaxPDVsPerPDU,
+                RequireMutualAuthentication = RequireMutualAuthentication
```
_RemoteCertificateValidationCallback_ is missing in Clone()

amoerie replied:
Nice catch. The tests work because this is set after the clone, but people wanting to use any of the Configure<TOptions> mechanisms to configure the default behavior would not be able to make this work. I'll fix this ASAP.

amoerie committed:
Also clone ServerCertificateValidationCallback + add docs

MBesuden commented on FO-DICOM.Core/Network/DesktopNetworkStreamFactory.cs:
```diff
@@ -0,0 +1,210 @@
+                var ssl = new SslStream(stream, false, userCertificateValidationCallback);
+                var sslHandshake =
+                    Task.Run(
+                        () => ssl.AuthenticateAsServerAsync(certificate, requireMutualAuthentication, SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12, false),
```
For the ClientStream the SslProtocols can be configured, for the Server this is hardcoded here.
For security reasons some older Tls protocol versions are already deprecated (currently recommended to use Tls 1.2 and higher).
Since there are already NetworkListenerCreationOptions used here maybe the SslProtocols could be configurable for the server as well.

MBesuden replied:
The last Parameter for the _checkCertificateRevocation_ could be given via NetworkListenerCreationOptions as well.

amoerie replied:
Your comment set me on a path to just separate the TLS parameters into separate options objects. 
It already felt slightly awkward to introduce server SSL options on DicomServiceOptions (which is also used by DicomClient!), so I've decided to take this a bit further and create new classes: DicomServerTlsOptions and DicomClientTlsOptions.

Please see my latest changes here: https://github.com/fo-dicom/fo-dicom/pull/1418/commits/47b233fba3eb268f8544628ae5a3d60f663a0101

amoerie committed:
Use dedicated options objects for server and client TLS

amoerie commented on FO-DICOM.Core/Network/DicomServer.cs:
```diff
@@ -274,7 +275,13 @@ private async Task ListenForConnectionsAsync()
+                {
+                    IpAddress = IPAddress,
+                    Port = Port,
+                    ServiceOptions = Options,
```
I wonder if this is still necessary

amoerie committed:
Remove ServiceOptions from NetworkListenerCreationOptions

This was no longer needed

amoerie committed:
Use TLS timeout for TLS handshake and regular timeout for network read/write

amoerie commented:
> I am not sure if the call ssl.AuthenticateAsClientAsync(options.Host) should be used anymore. Without client authentication (client certificates) it is not possible to specify which SslProtocols to use. Alternative: Use `ssl.AuthenticateAsClientAsync(options.Host, null, protocols, checkCertificateRevocation)` for that case (or unify both calls even when certificates is null or empty).

Does that work the way you suggest it works? I don't have time to test this right now.

amoerie commented:
> One thing I am not sure about is if the SslStream Properties SslStream.IsEncrypted and SslStream.IsSigned have to be checked after the ssl handshake...

amoerie committed:
Update changelog

MBesuden commented:
Sorry, i was checking the .Net 6 documentation where the _AuthenticateAsClientAsync_ call supports a nullable certificate collection.
Net 6: `public virtual System.Threading.Tasks.Task AuthenticateAsClientAsync (string targetHost, System.Security.Cryptography.X509Certificates.X509CertificateCollection? clientCertificates, System.Security.Authentication.SslProtocols enabledSslProtocols, bool checkCertificateRevocation);`

whereas in netstandard 2.0 at least an empty certificate collection is needed:
`public virtual System.Threading.Tasks.Task AuthenticateAsClientAsync (string targetHost, System.Security.Cryptography.X509Certificates.X509CertificateCollection clientCertificates, System.Security.Authentication.SslProtocols enabledSslProtocols, bool checkCertificateRevocation);`

No, i have not tested this before and the change is not necessarily needed.

amoerie committed:
Merge branch 'development' into GH-1402

amoerie committed:
Merge branch 'GH-1402' of github.com:fo-dicom/fo-dicom into GH-1402

amoerie committed:
Document breaking changes in release notes

amoerie committed:
Always use full overload of AuthenticateAsClientAsync

Based on https://github.com/microsoft/referencesource/blob/master/System/net/System/Net/SecureProtocols/SslStream.cs

        public virtual void AuthenticateAsClient(string targetHost)
        {
            AuthenticateAsClient(targetHost, new X509CertificateCollection(), ServicePointManager.DefaultSslProtocols, false);
        }

amoerie commented:
> Sorry, i was checking the .Net 6 documentation where the _AuthenticateAsClientAsync_ call supports a nullable certificate collection. Net 6: `public virtual System.Threading.Tasks.Task AuthenticateAsClientAsync (string targetHost, System.Security.Cryptography.X509Certificates.X509CertificateCollection? clientCertificates, System.Security.Authentication.SslProtocols enabledSslProtocols, bool checkCertificateRevocation);`
> 
> whereas in netstandard 2.0 at least an empty certificate collection is needed: `public virtual System.Threading.Tasks.Task AuthenticateAsClientAsync (string targetHost, System.Security.Cryptography.X509Certificates.X509CertificateCollection clientCertificates, System.Security.Authentication.SslProtocols enabledSslProtocols, bool checkCertificateRevocation);`
> 
> No, i have not tested this before and the change is not necessarily needed.

After perusing the SslStream source code, I have decided to add this change. Seems to work just fine. 👍

mrbean-bremen commented on ChangeLog5.md:
```diff
@@ -1,10 +1,14 @@
 #### 5.0.4 (TBD)
```
Even if we don't use semantic versioning, I would suggest to at least use version 5.1 as the next version, as a breaking change in a patch release is unexpected to most users.

amoerie replied:
I agree with this, but historically @gofal always decided on new version numbers. I wonder what his opinion is here.
If it's a minor patch bump, I also expect libraries to stay stable and not introduce breaking changes.

amoerie replied:
I'll resolve this convo and add a top level comment for @gofal

gofal replied:
sure, I am open for that. If there are some breaking changes or deeper changes with lot of internal refactorings a new minor version is fine.

mrbean-bremen commented on FO-DICOM.Core/Network/DicomServerTlsOptions.cs:
```diff
@@ -0,0 +1,44 @@
+    public class DicomServerTlsOptions
+    {
+        /// <summary>
+        /// The certificate the use for authenticated connections
```
the use -> to use

amoerie replied:
Done! Well spotted

mrbean-bremen commented on FO-DICOM.Core/Network/NetworkManager.cs:
```diff
@@ -13,27 +15,44 @@ public interface INetworkManager
         /// <param name="ipAddress">IP address(es) to listen to.</param>
         /// <param name="port">Network port to listen to.</param>
         /// <returns>Network listener implementation.</returns>
+        [Obsolete("Please use " + nameof(CreateNetworkListener) + "(" + nameof(NetworkListenerCreationOptions) + " options) instead")]
```
Maybe just a matter of taste - I find string interpolation ($"...") better readable.

amoerie replied:
I do too, but string interpolation is not available in C# Attributes because the strings have to be compile time constants. You can  try it for yourself, if it's not assignable to a private const string field, it's not usable as a parameter to an attribute.

mrbean-bremen replied:
Ah ok - thanks, didn't know this.

mrbean-bremen reviewed (approved):
I had checked this before and now mostly looked at the latest changes. It all looks good to me (with the usual caveat that I don't know the code good enough...) - thanks!

amoerie committed:
Fix wording of docs

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1402

amoerie committed:
Merge with development

MBesuden commented:
Hello,
Can be estimated when this feature will be completely reviewed and merged?

gofal committed:
Merge branch 'development' into GH-1402

amoerie commented:
> Hello, Can be estimated when this feature will be completely reviewed and merged?

It is awaiting review from the main contributor @gofal. After that, things can move fast.
Please be patient, everybody is a volunteer here.

amoerie committed:
Merge branch 'development' into GH-1402

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1402

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1402

mrbean-bremen commented:
@gofal - did you get the time to have a look? It has been a while...

gofal commented:
sure i had the time. there are two pending comments. One was, that the logger is a property of the class NetworkListenerCreationOptions, where the IP and Port of the listener are grouped together. And the other is, that there are some classes with nearly identical properties. The number of classes and factories and interfaces increased a lot with this PR. I wonder if it is necessary to have so many interfaces and factories, because it is getting more and more complicated which factory is creating which instance in which order. I used to have a good overview how the dicom services worked, but it took me a lot of effort to review this. For a open source project where we want people to participate and contribute it is important to keep code as simple, readable and understandable as possible. 
Amoeri is not active at the moment. So I will try to find some time, to rework the branch and try to find simpler solution.

amoerie commented:
> sure i had the time. there are two pending comments. One was, that the logger is a property of the class NetworkListenerCreationOptions, where the IP and Port of the listener are grouped together. And the other is, that there are some classes with nearly identical properties. The number of classes and factories and interfaces increased a lot with this PR. I wonder if it is necessary to have so many interfaces and factories, because it is getting more and more complicated which factory is creating which instance in which order. I used to have a good overview how the dicom services worked, but it took me a lot of effort to review this. For a open source project where we want people to participate and contribute it is important to keep code as simple, readable and understandable as possible. Amoeri is not active at the moment. So I will try to find some time, to rework the branch and try to find simpler solution.

That's valid feedback. I'll see if I can address some of your concerns.

amoerie committed:
Merge development

amoerie committed:
What if NetworkListenerCreationOptions did not have a logger

amoerie committed:
Then we can't pass a logger from DicomServer

amoerie committed:
Use injected logger in DesktopNetworkStreamFactory

amoerie committed:
Also remove logger from NetworkStreamCreationOptions

amoerie committed:
Remove redundant options in IDesktopNetworkStreamFactory.CreateAsServerAsync

amoerie commented:
I just reviewed my work, and this is a high level overview of what the network layer is doing during orchestration:

# Starting a server

A server goes through the following steps:

1. Create an instance of `IDicomServer` using `DicomServerFactory.Create`
2. When the server is created, it starts listening for incoming connections using `DicomServer.ListenForConnectionsAsync`
3. To do this, `DicomServer` calls `INetworkManager.CreateNetworkListener` to create an instance of `INetworkListener` (**simplification opportunity**: NetworkListener could be merged into DicomServer, but then you lose the ability to customize this via a custom NetworkManager)
4. `DicomServer` then calls  `INetworkListener.StartAsync` which starts a `TcpListener` under the hood
5. `DicomServer` then repeatedly accepts incoming connections via `INetworkListener.AcceptNetworkStreamAsync`, which triggers when a new `TcpClient` connects
6. `DesktopNetworkListener` calls `IDesktopNetworkStreamFactory.CreateAsServerAsync`,  
7. `IDesktopNetworkStreamFactory.CreateAsServerAsync` wraps the raw `NetworkStream` into a `DesktopNetworkStream`. It also handles the SSL handshake and timeouts. (simplification opportunity: `IDesktopNetworkStreamFactory.CreateAsServerAsync` could be merged into `INetworkListener`)
8. `DicomServer` then calls `DicomServer.CreateScp` to initialize a new `DicomService` using the `DesktopNetworkStream`
9. `DicomServer` then calls `DicomService.RunAsync` to begin handling incoming DICOM messages

# Starting a client

A client goes through the following steps:

1. Create an instance of `IDicomClient` using `DicomClientFactory.Create`
2. Add requests to this `IDicomClient` using `IDicomClient.AddRequestAsync` or `IDicomClient.AddRequestsAsync`
3. Begin sending using `IDicomClient.SendAsync`
4. `DicomClient` opens a connection using `IAdvancedDicomClientConnectionFactory.OpenConnectionAsync`
5. `IAdvancedDicomClientConnectionFactory.OpenConnectionAsync` creates a `NetworkStream` using `INetworkManager.CreateNetworkStream`
6. `INetworkManager.CreateNetworkStream` creates a `NetworkStream` using `IDesktopNetworkStreamFactory.CreateAsClientAsync`. (**Simplification opportunity**: `IDesktopNetworkStreamFactory.CreateAsClientAsync` could be merged into `NetworkManager`, or IAdvancedDicomClientConnectionFactory.OpenConnectionAsync` could call `IDesktopNetworkStreamFactory` directly. In the last case, we could remove some methods from NetworkManager)
7. `IDesktopNetworkStreamFactory.CreateAsClientAsync` creates a new `TcpClient`, opens a TCP connection and wraps the raw `NetworkStream` into a `DesktopNetworkStream`. It also handles the SSL handshake and timeouts.
8. When the `DesktopNetworkStream` is created, the connection is then used for the association handshake and the sending of the requests (outside the scope of this conversation)

# Simplification

So, there are multiple possible approaches that I want to present to you @gofal before actually simplifying/reducing the code.

## Possibility 1: remove NetworkManager (my preferred option)

I think we can remove `INetworkManager` as the middle man, because it contains increasingly less code, and we only maintain a single implementation anyway.

I would propose the following approach:

1. Replace `INetworkManager.CreateNetworkListener` with `INetworkListenerFactory.Create`. This factory could even be avoided and replaced with a simple constructor call, but then a fo-dicom consumer loses the ability to customize the TcpListener during creation.
2. Let `IAdvancedDicomClientConnectionFactory` call `IDesktopNetworkStreamFactory` directly
3. Remove remaining usages of `NetworkManager` (one question is what to do with `INetworkManager.IsSocketException`)
4. (Optional) Maybe rename DesktopNetworkListener -> NetworkListener, etc. You can perfectly use NetworkStream in Xamarin on mobile, so this whole "Desktop" naming scheme is maybe outdated.

## Possibility 2: remove DesktopNetworkStreamFactory

We can also remove `DesktopNetworkStreamFactory`, and move the two method implementations to `DesktopNetworkListener` and `DesktopNetworkManager` respectively. 
This is the least breaking change, but re-bloats the `NetworkManager` and doesn't really follow the single responsibility principle.
Furthermore, custom implementations of `INetworkManager` will probably inherit from `DesktopNetworkManager` and only override certain parts to avoid having to reimplement everything.

## Possibility 3: ???

So, my question is: do you agree with this analysis, and do you have a preference for one of the options or a suggestion for another option?
I am interested to hear your opinion @gofal, and also yours @mrbean-bremen

Side note: I have a tendency to create lots of little single responsibility classes which is simple to me, but I understand that for some people this increases the complexity instead. We will have to find some middle ground here. 😉

amoerie committed:
Correctly inject logger in DesktopNetworkStreamFactory

mrbean-bremen commented:
I'm fine with removing `INetworkManager`, as I don't see any need to re-implement the default implementation, but I may be missing some use case, of course. I have also no problem with many responsibility classes, but probably not all of them need to be exposed to the user. Anyway, your first option seems fine to me.

gofal commented:
Thanks for that very good overview. That helps much to understand how all that classes and factories work together.  I hope, we can simplify some of them.

First of all I completely agree, that the "Desktop" naming scheme can be omitted. This is all from version 4 where we had the shared project with the conrete implementations for desktop, andoroid, etc. But with latest .net version this all has been consilidated.

eg. there is the `INetworkManager` as interface with the abstract implementation `NetworkManager` and the (only) concrete implementation `DesktopNetworkManger`. I guess the last two classes can be merged without ony loss of flexibility.
The idea of the "*Manger" classes was to be a factory. If you take a look at the interface, then they mainly have some "Create" methods (and only few additional methods).
So when I then saw in this PR, that there is now also a new NetworkStreamFactory introduced, I noticed that this is the wrong direction. What about simply renaming the NetworkManger to NetworkFactory?
Then the method NetworkManger.CreatenetworkStream, which until now just created a new networkStream instance would become NetworkFactory.CreateNetworkStream, which also creates a Networkstream, but with a wrapped ssl stream if the parameter has the tlsOptions passed?

gofal commented:
Tomorrow I will look closer into the possibilities you mentiond.

gofal committed:
add unittest to verify unwrapped SocketException

amoerie commented:
> Thanks for that very good overview. That helps much to understand how all that classes and factories work together. I hope, we can simplify some of them.

It helped me review everything nicely too. :-) 

> First of all I completely agree, that the "Desktop" naming scheme can be omitted. This is all from version 4 where we had the shared project with the conrete implementations for desktop, andoroid, etc. But with latest .net version this all has been consilidated.

Nice! We can't just rename DesktopNetworkStream to NetworkStream though, that would be too confusing.
Perhaps DicomNetworkStream would be a good choice?
 
> eg. there is the `INetworkManager` as interface with the abstract implementation `NetworkManager` and the (only) concrete implementation `DesktopNetworkManger`. I guess the last two classes can be merged without ony loss of flexibility. The idea of the "*Manger" classes was to be a factory. If you take a look at the interface, then they mainly have some "Create" methods (and only few additional methods). So when I then saw in this PR, that there is now also a new NetworkStreamFactory introduced, I noticed that this is the wrong direction. What about simply renaming the NetworkManger to NetworkFactory? Then the method NetworkManger.CreatenetworkStream, which until now just created a new networkStream instance would become NetworkFactory.CreateNetworkStream, which also creates a Networkstream, but with a wrapped ssl stream if the parameter has the tlsOptions passed?

I agree with most of your points, but one detail is missing: you create a DesktopNetworkStream as a client or as a server. We can easily merge the client DesktopNetworkStream into NetworkManager and rename it to NetworkFactory, but that still leaves the DesktopNetworkStream as a server. Today, that is created in DesktopNetworkListener. Refactoring this class into NetworkFactory would be a bigger deal.

That is why my preference is to just remove NetworkManager. (Possibility 1 in my earlier post)

gofal committed:
Merge branch 'development' into GH-1402

amoerie committed:
Merge

amoerie committed:
Merge branch 'GH-1402' of github.com:fo-dicom/fo-dicom into GH-1402

amoerie commented:
> Tomorrow I will look closer into the possibilities you mentiond.

Hey @gofal have you had a chance to decide on this? 
Internally at my company we're eyeing to move our software to fo-dicom 5 and this is one of our major internal "customizations" to fo-dicom 4 that would become redundant, which would be nice.

amoerie committed:
Mark test as flaky in .NET Framework

gofal commented:
> Side note: I have a tendency to create lots of little single responsibility classes which is simple to me, but I understand that for some people this increases the complexity instead. We will have to find some middle ground here. 😉

"Single responsibility" is the keyword here. This PR should be about certificate handlig and now is about restructuring the network layers.

I therefore created a new PR #1514 to show a suggestion, how to solve the certificate issue.

According to paradigms like "Composition over interitance" and "KISS" it would be better to introduce new interfaces to deal with Tls instead of creating new layers in Factories.
So the change that you see there is, that simply instead of the boolean "UseTls" parameter in client constructor now there is a "ITlsInitiator" instance. And instead of the string "certificatename" parameter in server constructor now there is a "ITlsAcceptor" instance.
Those instances (that can be null in case of no Tls) will completely stand alone handle all the Tls stuff. So no additial layers in network classes is required. And all the options cerning Tls will not have to be passed through the whole layer stack.

and furthermore, with this interfaces, it should be more simple, to have a custom implementation based on either the windows certificate store, or based on pfx-files on a directory, or any other solution you want, without then having the need to completely implement the Networkmanager stuff.
And those instances then could be injected into DI and the factorymethods will take them by default if there is no explicit parameter.

gofal commented:
of course, simplifyint the networklayers, removing the "Desktop" naming schema, etc is a topic, but let's keep the PRs simple. They tend to mess up a lot.

gofal commented:
About simplifying the networklayers: as you wrote, this would be a bigger deal. Too big to add it here into an other topic, but fine enough to do it in a separate PR?

So I guess we all are in common, that the "DestopNetworkStream" as a wrapper around "Stream" would be fine, but renamed to "DicomNetworkStream". then there is still the flexibility of having the customized ConfigurableDesktopNetworkStreamDecorator in dicomClientTimeoutTests.

Currently the constructor of this DesktopNetworkStream/DicomNetworkStream is doing a lot of logic. this is not fine when this class should only be a wrapper, and it is also not so fine if we want to have async initialization logic (which is not possible within constructors).
So the most obvious thing would be, to have an DicomNetworkStreamFactory, that creates streams, handles tls stuff etc.
This DicomNetworkStreamFactory could then contain all the logic that now DesktopNetworkManger and NetworkManger (INetworkManager) contain. Since this is then a normal factory, this factory can be called whereever a new stream is created.

That can either be within the OpenConnection method of the AdvancedClientConnectionFactory,
or it can be within the AcceptNetworkStream of the DesktopNetworkListener (which by the way we also could rename to DicomNetworkListener).

I hope I have got the point of the previous posts correctly, then I think this change proposal should fit to all of them.
Or what did I miss? Any concerns?

amoerie commented:
Closing this PR in favor of #1514

amoerie commented:
> About simplifying the networklayers: as you wrote, this would be a bigger deal. Too big to add it here into an other topic, but fine enough to do it in a separate PR?
> 
> So I guess we all are in common, that the "DestopNetworkStream" as a wrapper around "Stream" would be fine, but renamed to "DicomNetworkStream". then there is still the flexibility of having the customized ConfigurableDesktopNetworkStreamDecorator in dicomClientTimeoutTests.
> 
> Currently the constructor of this DesktopNetworkStream/DicomNetworkStream is doing a lot of logic. this is not fine when this class should only be a wrapper, and it is also not so fine if we want to have async initialization logic (which is not possible within constructors).
> So the most obvious thing would be, to have an DicomNetworkStreamFactory, that creates streams, handles tls stuff etc.
> This DicomNetworkStreamFactory could then contain all the logic that now DesktopNetworkManger and NetworkManger (INetworkManager) contain. Since this is then a normal factory, this factory can be called whereever a new stream is created.
> 
> That can either be within the OpenConnection method of the AdvancedClientConnectionFactory,
> or it can be within the AcceptNetworkStream of the DesktopNetworkListener (which by the way we also could rename to DicomNetworkListener).
> 
> I hope I have got the point of the previous posts correctly, then I think this change proposal should fit to all of them.
> Or what did I miss? Any concerns?

I realize I never replied to this. I think you got about everything right. :-) 
I might be able to retrospectively do the refactoring in this branch and create a new pull request that only contains the refactoring. But I'll wait until your client certificates work has been merged.
